### PR TITLE
No boxing for structs

### PR DESCRIPTION
**Pull request by SergerGood**

### CODE REVIEW

File: src/EFCore/Query/CompiledQueryCacheKeyGenerator.cs [hunk 1]
```diff
@@ -121,10 +121,24 @@ public override bool Equals(object obj)
...
+            /// <returns>
+            ///     <see langword="true" /> if the current object is equal to the <paramref name="other" /> parameter; otherwise, <see langword="false" />.
+            /// </returns>
+            public bool Equals(CompiledQueryCacheKey other)
+            {
                 return ReferenceEquals(_model, other._model)
-                    && _queryTrackingBehavior == other._queryTrackingBehavior
-                    && _async == other._async
-                    && ExpressionEqualityComparer.Instance.Equals(_query, other._query);
+                   && _queryTrackingBehavior == other._queryTrackingBehavior
```
Comment (roji) [inline src/EFCore/Query/CompiledQueryCacheKeyGenerator.cs:139]: nit: 4 space indentation

File: src/EFCore/Query/CompiledQueryCacheKeyGenerator.cs [hunk 2]
```diff
@@ -76,7 +76,7 @@ public virtual object GenerateCacheKey(Expression query, bool async)
         ///         not used in application code.
         ///     </para>
         /// </summary>
-        protected readonly struct CompiledQueryCacheKey
+        protected readonly struct CompiledQueryCacheKey : IEquatable<CompiledQueryCacheKey>
```
Comment (smitpatel) [inline src/EFCore/Query/CompiledQueryCacheKeyGenerator.cs:79]: Also Relational/SqlServer and other provider specific cache keys

File: src/EFCore/Query/CompiledQueryCacheKeyGenerator.cs (hunk 1 above)
Comment (SergerGood) [reply src/EFCore/Query/CompiledQueryCacheKeyGenerator.cs]: Space indentation fixed and the Equals(object) method simplified

File: src/EFCore/Query/CompiledQueryCacheKeyGenerator.cs (hunk 2 above)
Comment (SergerGood) [reply src/EFCore/Query/CompiledQueryCacheKeyGenerator.cs:79]: OK, I will do it now

Comment (SergerGood) [reply src/EFCore/Query/CompiledQueryCacheKeyGenerator.cs:79]: Ready. I added what I found.

**Comment by roji** [comment]
Looks good, thanks! Two other structs which come to mind here are StoreObjectIdentifier (used as Dictionary key [here](https://github.com/dotnet/efcore/blob/7c572c5f570aff771586bf08c57aa73c424d4730/src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs#L236)), [ValueBuffer](https://github.com/dotnet/efcore/blob/2402ae49f52753262db1485bf3d5b68f43776c5b/src/EFCore/Storage/ValueBuffer.cs#L82) which already has Equals. @AndriySvyryd may have other ideas.

If you're uncomfortable with any of these of prefer to scope them out, just say so.

**Comment by SergerGood** [comment]
Oops) StoreObjectIdentifier added by someone. 
And ValueBuffer added in this commit already.

**Comment by roji** [comment]
Thanks for this contribution @SergerGood!

**Pull request merged**